### PR TITLE
avoid prepare_doc allocation

### PR DESCRIPTION
**Pull request by PSeitz**
avoid prepare_doc allocation, ~10% more thoughput best case

### CODE REVIEW

**Comment by adamreichold** [comment]
This is nice! Did you find this via allocation profiling? If so, which tool did you use?

**Comment by PSeitz** [comment]
> This is nice! Did you find this via allocation profiling? If so, which tool did you use?

@adamreichold I used perf `perf record --call-graph dwarf`. There was a relative large chunk of `free` in the calling function, so I was checking for allocations (`free` is still relative large though). It's also visible with an allocation profiler `valgrind --tool=dhat`.

File: src/schema/document.rs [hunk 1]
```diff
@@ -191,6 +191,34 @@ impl Document {
     pub fn get_first(&self, field: Field) -> Option<&Value> {
         self.get_all(field).next()
     }
+
+    /// Serializes stored field values.
+    pub fn serialize_stored<W: Write>(&self, schema: &Schema, writer: &mut W) -> io::Result<()> {
```
Comment (fulmicoton) [inline src/schema/document.rs:196]: can we make this a function in store/writer.rs?

**Review by fulmicoton — COMMENTED** [review]
Nice!

I have one blocking comment.

**Comment by fulmicoton** [comment]
@PSeitz Awesome. I love everything that removes allocation without makign the code too complex!
If you have creative ideas on how to remove the Doc related allocation that would be nice too.

I think we still have some in the tokenizer (T_T) and the document forces some too.
I think we can remove all of them.

File: src/schema/document.rs (hunk 1 above)
Comment (PSeitz) [reply src/schema/document.rs:196]: I think the abstraction is better like this, so the store doesn't need to know about `Document` internals like FieldValues and de/serialization is in the same file

Comment (fulmicoton) [reply src/schema/document.rs:196]: Ah yes you are right!

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**